From: SeongJae Park <sj@kernel.org>
Date: Thu, 27 Feb 2025 18:42:06 -0800
Subject: [PATCH] mm/damon/core: initialize damos->walk_completed in
 damon_new_scheme()

'damon_new_scheme()', which is a function for allocating and initialize
a 'struct damos' object, is not initializing ->walk_completed field.
The field is eventually set on damos_walk_complete(), but the first
damos_walk() could mistakenly not walk on the regions.  Actually, a
common usage of DAMOS for taking snapshot is installing a
monitoring-purpose DAMOS scheme, doing damos_walk() to retrieve the
snapshot, and then removing the scheme.  DAMON user-space tool (damo)
also gets runtime snapshot in the way.  Hence the problem can
continuously happen in such use cases.  Initialize it properly in the
allocation function.

Fixes: bf0eaba0ff9c ("mm/damon/core: implement damos_walk()")
Cc: <stable@vger.kernel.org> # 6.14.x
Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index 38f545fea585..cfa105ee9610 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -373,6 +373,7 @@ struct damos *damon_new_scheme(struct damos_access_pattern *pattern,
 	 * or damon_attrs are updated.
 	 */
 	scheme->next_apply_sis = 0;
+	scheme->walk_completed = false;
 	INIT_LIST_HEAD(&scheme->filters);
 	scheme->stat = (struct damos_stat){};
 	INIT_LIST_HEAD(&scheme->list);
-- 
2.39.5

